### DISCUSSION TITLE
Mac os big sur

### DISCUSSION THREAD

**Discussion by BokeMoire** [ImageMagick/ImageMagick#3357] [Help]
New operating system for me but an old problem.

mogrifying  -resize 1024x1024  /tmp/p340_1982.jpgqof4Mg.jpg
dyld: Library not loaded: /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib
  Referenced from: /Users/alan/ImageMagick-7.0.10/bin/mogrify

I solved this in the past by putting a link in the / directory.

On big sur, access to / is restricted and the kludge around it seems extremely convoluted.

Can the sources be download, the bad reference fixed, and compiled?

**Comment by urban-warrior** [comment]
We recommend Homebrew which provides pre-built binaries for Mac (some users prefer MacPorts). Download HomeBrew and type:
```
brew install imagemagick
```
Alternatively, build ImageMagick from source.

**Comment by BokeMoire** [comment]
Thanks for the reply.  
I don't want to use the brew stuff.

As I've used ImageMagick for years I've never looked around at other stuff.  This has forced me to so and NConvert appears to do what I want.

Shame.

**Reply by fmw42** [reply]
I do not know if this will help you, but I install all my delegates from MacPorts and then install ImageMagick from source.  If you think you want to go this way, let me know and I will provide links to what I have done.

**Reply by BokeMoire** [reply]
Thanks for the offer.  I am reluctant to use brew or macports.  I have found an alternative that, with initial testing, will do what I want.

The compiled in bad link has been unfixed for years and that clashed with the "we do not trust you with your own computer" changes on the mac.

Next laptop will probably be linux and I shall look at ImageMagick again.

**Comment by njdoyle** [comment]
If you're building from source, you probably need to actually do the install step to get the dylib's install name set properly.
Building with a prefix specified and installing should be sufficient:
```
./configure --prefix="/absolute/path/to/install/directory";
make;
make install;
```

**Reply by BokeMoire** [reply]
Thanks for the information.

Unfortunately, I have found another problem when trying to create another watermark picture using my Linux Computer.  It seems that -compose CopyOpacity doesn't work anymore.

I've solved the problem of applying watermarks using a program called nconvert but have yet to figure out how to create a new one.  It's strange as I actually found the original code for the watermark creation in the ImageMagick documentation.

**Reply by influencive** [reply]
> If you're building from source, you probably need to actually do the install step to get the dylib's install name set properly. Building with a prefix specified and installing should be sufficient:
> 
> ```
> ./configure --prefix="/absolute/path/to/install/directory";
> make;
> make install;
> ```

am also stuck on this issue how we fix this ?

**Comment by snibgo** [comment]
" -compose CopyOpacity -composite" works fine, as far as I know.

If it doesn't for you, please show your command and input files.

**Reply by BokeMoire** [reply]
I am getting very confused with this.  I have gone through it all again and CopyOpacity does appears to be working on my Linux computer.

I can create a new png watermark file on Linux, copy it my Mac and composite on the Mac is working for applying the watermark.

As I mentioned before, I'm have problems with mogrify and convert on the mac,
dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib
  Referenced from: /Users/alan/ImageMagick/bin/convert

because of the reference to the root directory which is not fixable with MacOS "big sur".

However, I am getting different results using the new watermark png instead of one of my old ones, when I use it with nconvert.  The old one is transparent but the new one has a white background.  Both work correctly with composite.  Which is a good result but confusing.

I am going to start again with the creation of new files on linux and investigate where I went wrong.  It looks like I should be able to use composite for the watermarks but I won't be able to resize with ImageMagick.

I may be back!

**Reply by fmw42** [reply]
Why can't you resize on ImageMagick?  If you are having transparency or composite issues, please post your input and command lines and we can try to help. Also post your IM version and exact platform/OS

**Reply by BokeMoire** [reply]
Because there is an absolute reference dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib

I have overcome this in the past by putting a symbolic link in the root directory.  But access to the root directory is not allowed in mac os big sur

**Reply by fmw42** [reply]
How did you install ImageMagick? Binary or source?  Where did you get your installation?

**Comment by BokeMoire** [comment]
Searching the internet seems to indicate that CopyOpacity disappeared from Imagemagick version 7, which I have on my Mac.  On my Linux computer, it is version 6.  This could be part of my problem.

I need to go through it all step by step.

**Reply by fmw42** [reply]
That is wrong. -compose copyOpacity is still there in IM 7.

**Comment by snibgo** [comment]
CopyOpacity works fine in v7. V7 syntax is stricter than v6, and that might be the problem.

Please show your command and input images.

**Comment by BokeMoire** [comment]
Starting at the beginning.

Below is a simpified bash script to run on my mac and my raspberry pi.  On the pi, it works and produces these files

pi_copyright_stamp.png  pi_stamp_fgnd.png  pi_stamp_mask.png

on the mac it produces

cellini:magictest alan$ ./copyright.sh 
dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib
  Referenced from: /Users/alan/ImageMagick/bin/convert
  Reason: image not found
./copyright.sh: line 13:  3956 Abort trap: 6           convert -size 650x50 xc:grey30 -font Times-Roman -pointsize 32 -gravity center -draw "scale 1,1 fill grey70  text 0,0  ${COPYRIGHT}" ${P}stamp_fgnd.png
dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib
  Referenced from: /Users/alan/ImageMagick/bin/convert
  Reason: image not found
./copyright.sh: line 18:  3957 Abort trap: 6           convert -size 650x50 xc:black -font Times-Roman -pointsize 32 -gravity center -draw "scale 1,1 fill white  text  1,1  ${COPYRIGHT}                               text  0,0  ${COPYRIGHT}                   fill black  text -1,-1 ${COPYRIGHT}" +matte ${P}stamp_mask.png
dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib
  Referenced from: /Users/alan/ImageMagick/bin/composite
  Reason: image not found
./copyright.sh: line 19:  3958 Abort trap: 6           composite -compose CopyOpacity ${P}stamp_mask.png ${P}stamp_fgnd.png ${P}copyright_stamp.png
dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib
  Referenced from: /Users/alan/ImageMagick/bin/mogrify
  Reason: image not found
./copyright.sh: line 20:  3959 Abort trap: 6           mogrify -trim +repage ${P}copyright_stamp.png
cellini:magictest alan$ 

Note that I have a symlink

lrwxr-xr-x  1 alan  staff  17  6 Mar 14:03 /Users/alan/ImageMagick@ -> ImageMagick-7.0.7

and

cellini:temp alan$ export
declare -x DYLD_LIBRARY_PATH="/Users/alan/ImageMagick/lib/"
declare -x MAGICK_HOME="/Users/alan/ImageMagick"


Here is the script

```
#!/bin/bash


COPYRIGHT="'© Alan H. Clifford'"
if [ `echo ${MACHTYPE} | grep -i "apple.*darwin"` ]  ; then
 P="mac_"
else
 P="pi_"
fi

convert -size 650x50 xc:grey30 -font Times-Roman -pointsize 32 -gravity center \
          -draw "scale 1,1 fill grey70  text 0,0  ${COPYRIGHT}" \
          ${P}stamp_fgnd.png
convert -size 650x50 xc:black -font Times-Roman -pointsize 32 -gravity center \
          -draw "scale 1,1 fill white  text  1,1  ${COPYRIGHT}  \
                             text  0,0  ${COPYRIGHT}  \
                 fill black  text -1,-1 ${COPYRIGHT}" \
          +matte ${P}stamp_mask.png
composite -compose CopyOpacity  ${P}stamp_mask.png  ${P}stamp_fgnd.png  ${P}copyright_stamp.png
mogrify -trim +repage ${P}copyright_stamp.png


```

**Reply by fmw42** [reply]
With IM 7, magick replace convert.  Do not use convert or magick convert. Just use magick.

**Comment by BokeMoire** [comment]
I downloaded the newer version of the mac binary release from https://imagemagick.org/script/download.php and changed my link and still get the same error with the path.

cellini:magictest alan$ ./copyright.sh 
dyld: Library not loaded: /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib
  Referenced from: /Users/alan/ImageMagick/bin/convert
  Reason: image not found
./copyright.sh: line 13:  4528 Abort trap: 6           convert -size 650x50 xc:grey30 -font Times-Roman -pointsize 32 -gravity center -draw "scale 1,1 fill grey70  text 0,0  ${COPYRIGHT}" ${P}stamp_fgnd.png
dyld: Library not loaded: /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib
  Referenced from: /Users/alan/ImageMagick/bin/convert
  Reason: image not found
./copyright.sh: line 18:  4529 Abort trap: 6           convert -size 650x50 xc:black -font Times-Roman -pointsize 32 -gravity center -draw "scale 1,1 fill white  text  1,1  ${COPYRIGHT}                               text  0,0  ${COPYRIGHT}                   fill black  text -1,-1 ${COPYRIGHT}" +matte ${P}stamp_mask.png
dyld: Library not loaded: /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib
  Referenced from: /Users/alan/ImageMagick/bin/composite
  Reason: image not found
./copyright.sh: line 19:  4530 Abort trap: 6           composite -compose CopyOpacity ${P}stamp_mask.png ${P}stamp_fgnd.png ${P}copyright_stamp.png
dyld: Library not loaded: /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib
  Referenced from: /Users/alan/ImageMagick/bin/mogrify
  Reason: image not found
./copyright.sh: line 20:  4531 Abort trap: 6           mogrify -trim +repage ${P}copyright_stamp.png
cellini:magictest alan$

**Reply by fmw42** [reply]
That binary is only good for High Sierra. What is your OSX version? If not High Sierra, then you should install from Homebrew.

**Reply by BokeMoire** [reply]
I don't really want to use homebrew but maybe I'll have a look.

**Reply by fmw42** [reply]
I am on a Mac and install all my delegates from MacPorts and then install IM from source. If interested, I can provide some links to how I do that.

**Reply by ivanooi** [reply]
> I am on a Mac and install all my delegates from MacPorts and then install IM from source. If interested, I can provide some links to how I do that.

Hi, do u mind to share the way you compile for Mac? is it static? is it under Mac 11? imagemagick 7.1.0 ? I'm having issues when copy the compiled binaries to another Mac. Compile under Mac 11 with ImageMagick 7.1.0 

hope we can share some info regarding static compilation. 
I had no issues if compile under Mac 10.10 with ImageMagick 7.0.10 

Thanks

**Reply by fmw42** [reply]
This is what I do on all my mac versions and all versions of IM 6 and 7.

```
./configure \
--enable-openmp  \
--enable-delegate-build --enable-shared --disable-static --disable-opencl \
--with-modules --with-quantum-depth=16 --without-wmf --with-rsvg --with-fftw \
--disable-silent-rules --disable-dependency-tracking --without-pango \
--with-lqr --with-gslib --with-gs-font-dir=/opt/local/share/ghostscript/fonts/

make clean
make

sudo make install
```

**Reply by ivanooi** [reply]
any reason you disabled the static build? thanks 

this is what my config : 

./configure \
    --disable-dependency-tracking \
    --disable-shared \
    --disable-delegate-build \
    --without-frozenpaths \
    --disable-installed \
    --disable-docs \
--prefix=/Users/ivanooi/Documents/IM-7.0.10-compiled2 \
--disable-shared \
    --with-webp \
    --with-raw \
--with-jpeg \
--with-png \
--with-tiff \
--with-wmf \
--with-rsvg \
--with-heic \
--disable-shared \
--with-modules

**Reply by fmw42** [reply]
It was recommended to do it that way years ago when I first started. I am not an expert on this and got help from the IM developers. So I cannot say why it was done that way.

**Reply by ivanooi** [reply]
I see... I'm trying to build a static IM but no luck for 7.1.0 under Mac 11.

**Comment by BokeMoire** [comment]
Next up is resize with

```
#!/bin/bash

COPYRIGHT="'© Alan H. Clifford'"
if [ `echo ${MACHTYPE} | grep -i "apple.*darwin"` ]  ; then
 P="mac_"
else
 P="pi_"
fi

cp big.jpg ${P}small.jpg
mogrifystr=" -resize 800x800 -quality 80 "
mogrify ${mogrifystr}  "${P}small.jpg"  # no quotes around mogrifystr

```

Works on the Pi.  On the mac I get

cellini:magictest alan$ ./resize.sh 
dyld: Library not loaded: /ImageMagick-7.0.7/lib/libMagickCore-7.Q16HDRI.5.dylib
  Referenced from: /Users/alan/ImageMagick/bin/mogrify
  Reason: image not found
./resize.sh: line 16:  5521 Abort trap: 6           mogrify ${mogrifystr} "${P}small.jpg"
cellini:magictest alan$

**Comment by snibgo** [comment]
I can't help with the dyld problem.

For IM v7, I would do it like this, in a single command:
```
magick \
  -size 650x50 xc:grey30 -font Times-Roman -pointsize 32 -gravity center \
  -draw "scale 1,1 fill grey70  text 0,0  COPYRIGHT" \
  -write stamp_fgnd.png \
  \( -size 650x50 xc:black -font Times-Roman -pointsize 32 -gravity center \
     -draw "scale 1,1 fill white  text  1,1  COPYRIGHT  \
                              text  0,0  COPYRIGHT  \
                  fill black  text -1,-1 COPYRIGHT" \
  \) \
  -alpha off \
  +write stamp_mask.png \
  -compose CopyOpacity -composite \
  -trim +repage \
  copyright_stamp.png
```
If you don't want to save intermediate results, you can remove "-write stamp_fgnd.png" and "-write stamp_mask.png".

**Comment by BokeMoire** [comment]
I might be making progress

cellini:magictest alan$ ./copyright.sh 
convert: unable to read font `Times-Roman' @ warning/annotate.c/RenderType/964.
convert: unable to read font `Times-Roman' @ error/annotate.c/RenderFreetype/1403.
convert: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4496.
convert: unable to read font `Times-Roman' @ warning/annotate.c/RenderType/964.
convert: unable to read font `Times-Roman' @ error/annotate.c/RenderFreetype/1403.
convert: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4496.
mogrify: geometry does not contain image `mac_copyright_stamp.png' @ warning/attribute.c/GetImageBoundingBox/510.
cellini:magictest alan$

**Reply by BokeMoire** [reply]
I did this

```
#!/bin/bash
# magick: set the correct path to libMagickCore.dylib 
install_name_tool -change \
    /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib \
    @executable_path/../lib/libMagickCore-7.Q16HDRI.8.dylib \
    /Users/alan/ImageMagick-7.0.10/bin/magick

# magick: set the correct path to libMagickWand.dylib 
install_name_tool -change \
    /ImageMagick-7.0.10/lib/libMagickWand-7.Q16HDRI.8.dylib \
    @executable_path/../lib/libMagickWand-7.Q16HDRI.8.dylib \
    /Users/alan/ImageMagick-7.0.10/bin/magick

# libMagickWand.dylib: set the correct ID
install_name_tool -id \
    @executable_path/../lib/libMagickWand-7.Q16HDRI.8.dylib \
    /Users/alan/ImageMagick-7.0.10/lib/libMagickWand-7.Q16HDRI.8.dylib

# libMagickWand.dylib: set the correct path
install_name_tool -change \
    /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib \
    @executable_path/../lib/libMagickCore-7.Q16HDRI.8.dylib \
    /Users/alan/ImageMagick-7.0.10/lib/libMagickWand-7.Q16HDRI.8.dylib

# libMagickCore.dylib: set the correct ID
install_name_tool -id \
    @executable_path/../lib/libMagickCore-7.Q16HDRI.8.dylib \
    /Users/alan/ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib

```

and installed Xquartz

**Reply by fmw42** [reply]
On IM 7, use magick in place of convert. Also preface all other tools with magick, such as, magick identify and magick mogrify, etc.

**Reply by BokeMoire** [reply]
The commands are links to magick in the bin directory

**Reply by chern041** [reply]
This works for me perfectly, thank you!

**Comment by BokeMoire** [comment]
I've produced a watermark using a direct path to a font. Not how it should be tboug

```
convert -size 650x50 xc:grey30 -font /System/Library/Fonts/Helvetica.ttc -pointsize 32 -gravity cente$
          -draw "scale 1,1 fill grey70  text 0,0  ${COPYRIGHT}" \
          ${P}stamp_fgnd.png
convert -size 650x50 xc:black -font /System/Library/Fonts/Helvetica.ttc -pointsize 32 -gravity center$
          -draw "scale 1,1 fill white  text  1,1  ${COPYRIGHT}  \
                             text  0,0  ${COPYRIGHT}  \
                 fill black  text -1,-1 ${COPYRIGHT}" \
          +matte ${P}stamp_mask.png
composite -compose CopyOpacity  ${P}stamp_mask.png  ${P}stamp_fgnd.png  ${P}copyright_stamp.png
mogrify -trim +repage ${P}copyright_stamp.png

```



![image](https://user-images.githubusercontent.com/52888734/110670270-1f622c00-81c5-11eb-8127-ec73bb468356.png)

**Reply by fmw42** [reply]
See the script, imagick_type_gen, at https://legacy.imagemagick.org/Usage/scripts/  It searches your system for fonts and creates a type.xml file so that you can reference fonts by name rather than path to the font.

**Reply by BokeMoire** [reply]
Thanks, looking at that now.

**Reply by BokeMoire** [reply]
It needed a little modification

`find /System/Library/Fonts -type f -name '*.*' | ./imagick_type_gen -f - > type.xml`

but I've got an xml file.

**Reply by BokeMoire** [reply]
Another path problem

magick -list font

only finds the file if it is in the current directory.  It is not finding ~/.magick

... later

moved type.xml to /Users/alan/ImageMagick/etc/ImageMagick-7/

**Comment by BokeMoire** [comment]
Almost managed it.  The background of the watermark should be transparent.

![image](https://user-images.githubusercontent.com/52888734/110671092-0a39cd00-81c6-11eb-96f4-bbc27ca2b77d.png)

**Comment by snibgo** [comment]
Your commands work fine for me with v6.9.9-50. For v7, I suggest you use "magick". If you use "convert" etc then you get v6 syntax and v7 processing. This is a temporary workaround only for scripts that can't be adapted.

For all new development, I suggest using "magick".

See my example command above.

**Reply by BokeMoire** [reply]
With 

```
magick \
  -size 650x50 xc:grey30 -font /System/Library/Fonts/Times.ttc -pointsize 32 -gravity center \
  -draw "scale 1,1 fill grey70  text 0,0  COPYRIGHT" \
  -write stamp_fgnd.png \
  \( -size 650x50 xc:black -font /System/Library/Fonts/Times.ttc -pointsize 32 -gravity center \
     -draw "scale 1,1 fill white  text  1,1  COPYRIGHT  \
                              text  0,0  COPYRIGHT  \
                  fill black  text -1,-1 COPYRIGHT" \
  -alpha off \
  +write stamp_mask.png \
  -compose CopyOpacity -composite \
  -trim +repage \
  copyright_stamp7.png
```


I get an error

magick: image sequence is required `-composite' @ error/operation.c/CLIListOperatorImages/4000.


This works like the original

```
magick convert -size 650x50 xc:grey30  -font /System/Library/Fonts/Helvetica.ttc -pointsize 32 -gravity center \
          -draw "scale 1,1 fill grey70  text 0,0  ${COPYRIGHT}" \
          ${P}stamp_fgnd.png
magick convert -size 650x50 xc:black -font /System/Library/Fonts/Helvetica.ttc  -pointsize 32 -gravity center \
          -draw "scale 1,1 fill white  text  1,1  ${COPYRIGHT}  \
                             text  0,0  ${COPYRIGHT}  \
                 fill black  text -1,-1 ${COPYRIGHT}" \
          +matte ${P}stamp_mask.png
magick composite -compose CopyOpacity  ${P}stamp_mask.png  ${P}stamp_fgnd.png  ${P}copyright_stamp.png
magick mogrify -trim +repage ${P}copyright_stamp.png

```

**Reply by fmw42** [reply]
Don't use magick convert. On IM 7, use magick in place of convert.

**Comment by snibgo** [comment]
I corrected my command, after omitting the close parenthesis: )

**Reply by BokeMoire** [reply]
Looking good.  Thanks

**Comment by BokeMoire** [comment]
Thanks for the help.  

ImageMagick seems to be doing what I want on Mac OS big sur.  I can create a couple of jpegs, black and white, resize them, create a watermark, put the watermark on the small jpegs.

http://clifford.ac/~alan/doneitwhite.jpg
http://clifford.ac/~alan/doneitblack.jpg

Now to make sure it re-integrates with my photo processing scripts.

**Comment by BokeMoire** [comment]
ivanooi wrote to ImageMagick/ImageMagick
[at 18:16 (-0000) on Tuesday, 8th March, 2022]:

> I see... I'm trying to build a static IM but no luck for 7.1.0 under Mac 11.
>
>

These are the notes I took

ImageMagick on Mac OS Big Sur

Install ImageMagick-x86_64-apple-darwin20.1.0.tar.gz 
from https://imagemagick.org/script/download.php
-----------------------------------------------------------

Put a symbolic link in home

/Users/alan/ImageMagick@ -> ImageMagick-7.0.10
-----------------------------------------------------------

Add /Users/alan/ImageMagick/bin to the path
-----------------------------------------------------------

Apparently

export DYLD_LIBRARY_PATH="/Users/alan/ImageMagick/lib/"

is ignored on the mac and doesn't become part of the environment.
-----------------------------------------------------------


Modify the following for current names and run

#!/bin/bash
# magick: set the correct path to libMagickCore.dylib 
install_name_tool -change \
     /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib \
     @executable_path/../lib/libMagickCore-7.Q16HDRI.8.dylib \
     /Users/alan/ImageMagick-7.0.10/bin/magick

# magick: set the correct path to libMagickWand.dylib 
install_name_tool -change \
     /ImageMagick-7.0.10/lib/libMagickWand-7.Q16HDRI.8.dylib \
     @executable_path/../lib/libMagickWand-7.Q16HDRI.8.dylib \
     /Users/alan/ImageMagick-7.0.10/bin/magick

# libMagickWand.dylib: set the correct ID
install_name_tool -id \
     @executable_path/../lib/libMagickWand-7.Q16HDRI.8.dylib \
     /Users/alan/ImageMagick-7.0.10/lib/libMagickWand-7.Q16HDRI.8.dylib

# libMagickWand.dylib: set the correct path
install_name_tool -change \
     /ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib \
     @executable_path/../lib/libMagickCore-7.Q16HDRI.8.dylib \
     /Users/alan/ImageMagick-7.0.10/lib/libMagickWand-7.Q16HDRI.8.dylib

# libMagickCore.dylib: set the correct ID
install_name_tool -id \
     @executable_path/../lib/libMagickCore-7.Q16HDRI.8.dylib \
     /Users/alan/ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.8.dylib
-----------------------------------------------------------


Install Xquartz
-----------------------------------------------------------


From here https://legacy.imagemagick.org/Usage/scripts/
get the imagick_type_gen script and then run

find /System/Library/Fonts -type f -name '*.*' | ./imagick_type_gen -f - > type.xml

put type.xml into

~/ImageMagick/etc/ImageMagick-7/type.xml

-----------------------------------------------------------

Use magic 7 terminology such as

magick convert

instead of

convert

I need to find out more about this.

-----------------------------------------------------------
